src/mlpack/tests/ksinit_test.cpp

```diff
@@ -0,0 +1,315 @@
+    validationErrorAvg += validationError;
+
+  }
+
+  trainErrorAvg /= k;
+  validationErrorAvg  /= k;
+
+  Error e;
+
+  e.trainError = trainErrorAvg;
+  e.validationError = validationErrorAvg;
+
+  return e;
+
+}
+
+/*
+  AvgCrossValidation function takes a dataset and runs CrossValidation "iter"
+  number of times and then return the average training and validation error.
+  It shuffles the dataset in every iteration.
+
+*/
+
+Error AvgCrossValidation(arma::mat& dataset, size_t numLabels, size_t iter,
+                         const size_t hiddenLayerSize, const size_t maxEpochs)
+{
+
+/*
+  @params dataset The dataset inclusive of the labels. Assuming the last 
+                  "numLabels" number of rows are the labels which the model
+                  has to predict. 
+  @params numLabels number of rows which are the output labels in the dataset.
+
+  @params iter The number of times Cross Validation has to be run.
+
+  @params hiddenLayerSize The number of nodes in the hidden layer.
+
+  @params maxEpochs The maximum number of epochs for the training.
+
+  @params avgError Returns the train and validation error averaged over "iter"
+                   number of Cross Validation results.
+
+*/
+  Error avgError = {0.0, 0.0};
+
+  for (size_t i = 0; i < iter;)
+  {
+    dataset = arma::shuffle(dataset, 1);
+
+    arma::mat trainData = dataset.submat(0, 0, dataset.n_rows - 1 - numLabels, 
+                                         dataset.n_cols - 1);
+    arma::mat trainLabels = dataset.submat(dataset.n_rows - numLabels, 0, 
+                                           dataset.n_rows - 1, 
+                                           dataset.n_cols - 1); 
+    Error e = CrossValidation(trainData, trainLabels, 10, hiddenLayerSize, 
+                              maxEpochs);
+
+    // The check is to prevent nan values from affecting the result.
+    if (arma::is_finite(e.trainError) && arma::is_finite(e.validationError))
+    {
+      avgError.trainError += e.trainError;
+      avgError.validationError += e.validationError;
+      ++i;
+    }
+    
+  }
+
+  avgError.trainError /= iter;
+  avgError.validationError /= iter;
+
+  return avgError;
+}
+
+
+/*Test case for the Iris Dataset */
+
+BOOST_AUTO_TEST_CASE(IrisDataset)
+{
+  arma::arma_rng::set_seed_random();
+
+  double trainErrorThreshold = 0.001;
+  double validationErrorThreshold = 0.001;
+
+  arma::mat dataset, labels;
+
+  data::Load("iris.csv", dataset, true);
+  data::Load("iris_labels.txt", labels, true);
+
+  dataset.insert_rows(dataset.n_rows, labels);
+
+  dataset /= 10; // Normalization used in the paper.
+
+  Error avgError = AvgCrossValidation(dataset, 1, 10, 3, 52); //Runs CV 10 times
+
+  BOOST_REQUIRE_LE(avgError.trainError, trainErrorThreshold);
+  BOOST_REQUIRE_LE(avgError.validationError, validationErrorThreshold);
+
+}
+
+// Test case for the Non Linear Function Approximation Problem.
+
+BOOST_AUTO_TEST_CASE(NonLinearFunctionApproximation)
```

Comment:
I have completely forgotten the non linear function approximation task, looks nice. Thanks for writing another test case :+1: